### ISSUE TITLE
How to tune speed loop PID parameters?

### ISSUE THREAD

**Issue by cnpcshangbo** [cnpcshangbo/simscapeFOPD#5]
Hi everyone?
I am having difficulties tuning the speed loop PID parameters. I followed https://github.com/cnpcshangbo/simscapeFOPD/issues/2 to get a speed curve and then I changed the speed setpoint to step function according to this: https://github.com/cnpcshangbo/simscapeFOPD/issues/4. After that I was able to get speed loop step response like this:
![image](https://user-images.githubusercontent.com/4831029/36654010-9b1b7bb6-1af4-11e8-9e49-c1988d8723c4.png)
Speed loop P=2, I=6, D=0.324786114614273, N=19.7366420228035, ITAE=2.35

**Comment by cnpcshangbo** [comment]
From the figure above I think the P parameter should be reduced because the overshoot is too large, So let's cut it to half and see.
Speed loop P=1, I=6, D=0.324786114614273, N=19.7366420228035, ITAE=3.66.

![image](https://user-images.githubusercontent.com/4831029/36654287-76937882-1af6-11e8-93b4-8d28aff65027.png)

**Comment by cnpcshangbo** [comment]
Since getting the P smaller makes the curve worse. I am going to make P larger.
Speed loop P=4, I=6, D=0.324786114614273, N=19.7366420228035, ITAE=1.95.

![image](https://user-images.githubusercontent.com/4831029/36654380-2cdd63c8-1af7-11e8-8efc-f3bd8b658bd5.png)

**Comment by cnpcshangbo** [comment]
Getting P larger really works. It gets overshoot smaller, so let try to get P even larger.
P=8, I=6, D=0.324786114614273, N=19.7366420228035, ITAE=1.66.

![image](https://user-images.githubusercontent.com/4831029/36654503-13accf32-1af8-11e8-86ad-6b7ed708e059.png)

**Comment by cnpcshangbo** [comment]
P = 8 is really better. Since we already got some oscillation, we shouldn't get P larger.
We'll try to get integrator larger and see what will happen.
P=8, I=10, D=0.324786114614273, N=19.7366420228035, ITAE=1.78
![image](https://user-images.githubusercontent.com/4831029/36654606-cb58a732-1af8-11e8-90d5-6177fc27ecbb.png)

**Comment by cnpcshangbo** [comment]
Since increasing I makes the system worse, we'll try to decrease I.
P=8, I=3, D=0.324786114614273, N=19.7366420228035, ITAE=2.35
![image](https://user-images.githubusercontent.com/4831029/36654693-5a272394-1af9-11e8-8a6b-117cf0eba7fe.png)

**Comment by cnpcshangbo** [comment]
I=3 makes the overshoot less, but it also makes the static error converge slower. So we think this is the best we can get:
P=8, I=6, D=0.324786114614273, N=19.7366420228035, ITAE=1.66
![image](https://user-images.githubusercontent.com/4831029/36654841-03d7b4ee-1afa-11e8-8131-4bed1e130be4.png)

**Comment by cnpcshangbo** [comment]
I think the speed loop is much better than before. Therefore I want to test whether IOPID and FOPD is getting better.

**Comment by cnpcshangbo** [comment]
![image](https://user-images.githubusercontent.com/4831029/36655760-9abd6534-1aff-11e8-90b7-871bec09edb1.png)
ZPos PID continous P=6.25, I=4.81, D=1.81 anti-windup is used. Kb=1.
The altitude ITAE is 1.13 for IOPID controller. I am going to tune the altitude loop now.
I have created a new issue for the altitude loop tuning.
https://github.com/cnpcshangbo/simscapeFOPD/issues/7

**Comment by cnpcshangbo** [comment]
![image](https://user-images.githubusercontent.com/4831029/36656052-5586493e-1b01-11e8-9171-0427213ab7aa.png)
ZPos PID Discrete-time P=3.2, I=2, D=0, no N, ITAE=0.99
Kb=1, output limit [-10, 10]

**Comment by cnpcshangbo** [comment]
The next step is to re-identify the plant model for ZPos(altitude) control.
I plan to finish that by using the PID tune APP (refer to the webinar from Mathworks).

**Comment by cnpcshangbo** [comment]
![image](https://user-images.githubusercontent.com/4831029/36656637-44264876-1b04-11e8-97a9-de1d7d9726f7.png)
So the ZPos plant transfer function is: 0.99217/s/(0.2486*s+1).

**Comment by cnpcshangbo** [comment]
![image](https://user-images.githubusercontent.com/4831029/36657291-f853f0f8-1b06-11e8-8221-8efdda1f8911.png)
PID Tuner is not able to tune this plant. So I plan to tune it using my code from FOPD tunner.

**Comment by cnpcshangbo** [comment]
By using this project:
https://github.com/cnpcshangbo/FOPD-tunner
I am able to get the optimal IOPID parameters.
C =
 
             1          
  Kp + Ki * --- + Kd * s
             s          

  with Kp = 8.55, Ki = 8.79, Kd = 1.85
 
Continuous-time PID controller in parallel form.


info = 

  struct:

                Stable: 1
    CrossoverFrequency: 7.1642
           PhaseMargin: 83.8783

![image](https://user-images.githubusercontent.com/4831029/36657488-cd4284dc-1b07-11e8-853e-6995cd6ae641.png)
